src/grammars/syntax.md

```diff
@@ -1,5 +1,453 @@
+
+This leads to the common idiom meaning "any character but":
+
+```
+not_space_or_tab = {
+    !(                // if the following text is not
+        " "           //     a space
+        | "\t"        //     or a tab
+    )
+    ~ any             // then consume one character
+}
+
+triple_quoted_string = {
+    "'''"
+    ~ triple_quoted_character*
+    ~ "'''"
+}
+triple_quoted_character = {
+    !"'''"        // if the following text is not three apostrophes
+    ~ any         // then consume one character
+}
+```
+
+## Operator precedence and grouping (WIP)
+
+The repetition operators asterisk `*`, plus sign `+`, and question mark `?`
+apply to the immediately preceding expression.
+
+```
+"One " ~ "or " ~ "more. "+
+"One " ~ "or " ~ ("more. "+)
+    are equivalent and match
+"One or more. more. more. more. "
+```
+
+Larger expressions can be repeated by surrounding them with parentheses.
+
+```
+("One " ~ "or " ~ "more. ")+
+    matches
+"One or more. One or more. "
+```
+
+Repetition operators have the highest precedence, followed by predicate
+operators, the sequence operator, and finally ordered choice.
+
+```
+my_rule = {
+    "a"* ~ "b"?
+    | &"b"+ ~ "a"
+}
+    equivalent to
+my_rule = {
+      ( ("a"*) ~ ("b"?) )
+    | ( (&("b"+")) ~ "a" )
+}
+```
+
+## Start and end of input
+
+The rules `soi` and `eoi` match the *start* and *end* of the input string,
+respectively. Neither consumes any text. They only indicate whether the parser
+is currently at one edge of the input.
+
+For example, to ensure that a rule matches the entire input, where any syntax
+error results in a failed parse (rather than a successful but incomplete
+parse):
+
+```
+main = {
+    soi
+    ~ (...)
+    ~ eoi
+}
+```
+
+## Implicit whitespace
+
+Many languages and text formats allow arbitrary whitespace and comments between
+logical tokens. For instance, Rust considers `4+5` equivalent to `4 + 5` and `4
+/* comment */ + 5`.
+
+The **optional rules `whitespace` and `comment`** implement this behaviour. If
+either (or both) are defined, they will be implicitly inserted at every
+[sequence] and between every [repetition] (except in [atomic rules]).
+
+```
+expression = { "4" ~ "+" ~ "5" }
+whitespace = _{ " " }
+comment = _{ "/*" ~ (!"*/" ~ any)* ~ "*/" }
+    matches
+"4+5"
+"4 + 5"
+"4  +     5"
+"4 /* comment */ + 5"
+```
+
+As you can see, `whitespace` and `comment` are run repeatedly, so they need
+only match a single whitespace character or a single comment. The grammar above
+is equivalent to:
+
+```
+expression = {
+    "4"   ~ (ws | com)*
+    ~ "+" ~ (ws | com)*
+    ~ "5"
+}
+ws = _{ " " }
+com = _{ "/*" ~ (!"*/" ~ any)* ~ "*/" }
+```
+
+(Be sure to mark the `whitespace` and `comment` rules as [silent] unless you
+want to see them included inside other rules!)
+
+[sequence]: #sequence
+[repetition]: #repetition
+[atomic rules]: #atomic
+[silent]: #silent-and-atomic-rules
+
+## Silent and atomic rules
```

Comment:
It might be worth mentioning how these modifiers influence error reporting, but maybe that's better left for an ulterior chapter?

Reply:
Error reporting had slipped my mind, so thanks for bringing it up! It probably deserves its own subchapter.